docs_nnx/guides/transforms.md

```diff
@@ -0,0 +1,340 @@
+  y = x @ weights.kernel + weights.bias
+  weights.some_property = ['a', 2, False] # add attribute
+  del weights.bias # delete attribute
+  weights.new_param = weights.kernel # share reference
+  return y
+
+y = nnx.vmap(crazy_vector_dot, in_axes=0, out_axes=1)(weights, x)
+
+nnx.display(weights)
+```
+
+> With great power comes great responsibility.
+> <br> \- Uncle Ben
+
+While this feature is very powerful, it must be used with care as it can clash with JAX's underlying assumptions for certain transformations. For example, `jit` expects the structure of the inputs to be stable in order to cache the compiled function, changing the graph structure inside a `nnx.jit`-ed function cause continuous recompilations and performance degradation, `scan` on the other hand only allows a fixed `carry` structure, so adding/removing substates declared as carry will cause an error.
+
++++
+
+## Transforming Substates (Lift Types)
+
+Certain JAX transformation allow the use of pytree prefixes to specify how different parts of the inputs/outputs should be transformed. Flax NNX supports pytree prefixes for pytree structures but currently it doesn't have the notion of a prefix for graph objects. Instead, Flax NNX introduces the concept of `Lift Types` which allow specifying how different substates of an object should be transformed. Different transformations support different Lift Types, here is the list of currently supported Lift Types for each transformation:
+
+| Lift Type        | Transforms                              |
+|------------------|-----------------------------------------|
+| `StateAxes`      | `vmap`, `pmap`, `scan`                  |
+| `StateSharding`  | `jit`, `shard_map`                      |
+| `DiffState`      | `grad`, `value_and_grad`, `custom_vjp`  |
+
+> NOTE: `shard_map` is not yet implemented.
+
+If we want to specify how to vectorize different substates of an object in `nnx.vmap`, we create a `StateAxes` which maps a set of substates via [Filters](https://flax-nnx.readthedocs.io/en/latest/guides/filters_guide.html) to their corresponding axes, and pass the `StateAxes` to `in_axes` and `out_axes` as if it were a pytree prefix. Let's use the previous `stateful_vector_dot` example and
+vectorize only the `Param` variables and broadcast the `count` variable so we only keep a single count for all the batch elements.
+To do this we will define a `StateAxes` with a filter that matches the `Param` variables and maps them to axis `0`, and all the `Count` variables to `None`, and pass this `StateAxes` to `in_axes` for the `Weights` object.
+
+```{code-cell} ipython3
+class Weights(nnx.Module):
+  def __init__(self, kernel: jax.Array, bias: jax.Array, count: jax.Array):
+    self.kernel, self.bias = nnx.Param(kernel), nnx.Param(bias)
+    self.count = Count(count)
+
+weights = Weights(
+  kernel=random.uniform(random.key(0), (10, 2, 3)),
+  bias=jnp.zeros((10, 3)),
+  count=jnp.array(0),
+)
+x = jax.random.normal(random.key(1), (10, 2))
+
+
```

Comment:
It's probably better to only call vmap once when only one call is needed, to avoid confusion. Same for the example below.

```
state_axes = nnx.StateAxes({nnx.Param: 0, Count: None}) # broadcast Count
@nnx.vmap(in_axes=(state_axes, 0), out_axes=1)
def stateful_vector_dot(weights: Weights, x: jax.Array):
  assert weights.kernel.ndim == 2, 'Batch dimensions not allowed'
  assert x.ndim == 1, 'Batch dimensions not allowed'
  weights.count += 1
  return x @ weights.kernel + weights.bias

y = stateful_vector_dot(weights, x)
y = stateful_vector_dot(weights, x)
```